source/slang/slang-check-expr.cpp

```diff
@@ -1355,6 +1355,162 @@ namespace Slang
         }
     }
 
+    RefPtr<Expr> SemanticsVisitor::CheckMatrixSwizzleExpr(
+        MemberExpr* memberRefExpr,
+        RefPtr<Type>      baseElementType,
+        IntegerLiteralValue baseElementRowCount,
+        IntegerLiteralValue baseElementColCount)
+    {
+        RefPtr<MatrixSwizzleExpr> swizExpr = new MatrixSwizzleExpr();
+        swizExpr->loc = memberRefExpr->loc;
+        swizExpr->base = memberRefExpr->baseExpression;
+
+        int elementIndices[4][2];
+        int elementCount = 0;
+
+        bool anyDuplicates = false;
+        int oneIndexed = -1;
+
+        String swizzleText = getText(memberRefExpr->name);
+        auto coursor = swizzleText.begin();
+
+        while (coursor != swizzleText.end())
+        {
+            // Throw out swizzling with more than 4 output elements
+            if (elementCount >= 4) {
+                getSink()->diagnose(swizExpr, Diagnostics::invalidSwizzleExpr, swizzleText, baseElementType->toString());
+                return CreateErrorExpr(memberRefExpr);
+            }
+            int elementIndex[2];
+            elementIndex[0] = 0;
+            elementIndex[1] = 0;
+            // Check for the preceding underscore
+            if (*coursor++ != '_') {
+                getSink()->diagnose(swizExpr, Diagnostics::invalidSwizzleExpr, swizzleText, baseElementType->toString());
+                return CreateErrorExpr(memberRefExpr);
+            }
+
+            // Check for one or zero indexing            
+            if (*coursor == 'm') {
+                // Can't mix one and zero indexing
+                if (oneIndexed == 1) {
+                    getSink()->diagnose(swizExpr, Diagnostics::invalidSwizzleExpr, swizzleText, baseElementType->toString());
```

Comment:
Nit pick: A message describing the problem here a little more precisely might be nice.